rust/src/measurements/make_private_expr/expr_count/mod.rs

```diff
@@ -0,0 +1,145 @@
+use crate::domains::MarginPub;
+use crate::metrics::PartitionDistance;
+use crate::transformations::traits::UnboundedMetric;
+use crate::transformations::StableExpr;
+use crate::{
+    core::{Function, Measurement},
+    domains::ExprDomain,
+    error::Fallible,
+};
+
+use num::Zero;
+use polars::lazy::dsl::Expr;
+use polars_plan::dsl::AggExpr;
+
+/// Make a measurement that computes the count exactly under bounded-DP
+///
+/// | input_metric                              |
+/// | ----------------------------------------- |
+/// | `PartitionDistance<SymmetricDistance>`    |
+/// | `PartitionDistance<InsertDeleteDistance>` |
+///
+/// # Arguments
+/// * `input_domain` - ExprDomain
+/// * `input_metric` - valid selections shown in table above
+/// * `expr` - count expression
+pub fn make_expr_private_count<MI: 'static + UnboundedMetric, MO: 'static + Measure>(
+    input_domain: ExprDomain,
+    input_metric: PartitionDistance<MI>,
+    expr: Expr,
+) -> Fallible<Measurement<ExprDomain, Expr, PartitionDistance<MI>, MO>>
+where
+    MO::Distance: Zero,
+    Expr: StableExpr<PartitionDistance<MI>, PartitionDistance<MI>>,
+    (ExprDomain, PartitionDistance<MI>): MetricSpace,
+{
+    let Expr::Agg(AggExpr::Count(expr, include_nulls)) = expr else {
+        return fallible!(MakeMeasurement, "Expected Count expression");
+    };
+
+    if !include_nulls {
+        return fallible!(
+            MakeMeasurement,
+            "Public margin sizes assume nulls are included. Try using `len` instead."
+        );
+    }
+
+    let t_prior = expr
+        .clone()
+        .make_stable(input_domain.clone(), input_metric.clone())?;
+
+    let (middle_domain, middle_metric) = t_prior.output_space();
+
+    let by = middle_domain.context.grouping_columns()?;
+    let margin = middle_domain
+        .frame_domain
+        .margins
+        .get(&by)
+        .ok_or_else(|| err!(MakeMeasurement, "Unknown margin for {:?}", by))?;
+
+    if Some(MarginPub::Lengths) != margin.public_info {
+        return fallible!(
+            MakeMeasurement,
+            "Len is only private if size(s) of {:?} are public",
+            by
+        );
+    }
+
+    t_prior
+        >> Measurement::new(
+            middle_domain,
+            Function::new_expr(|input_expr| input_expr.len()),
+            middle_metric,
+            MO::default(),
+            PrivacyMap::new(move |_| MO::Distance::zero()),
+        )?
+}
+
+#[cfg(test)]
+mod test_make_expr_count_private {
+    use super::*;
+    use polars::prelude::*;
+
+    use crate::{
+        error::ErrorVariant,
+        measurements::PrivateExpr,
+        measures::MaxDivergence,
+        metrics::{PartitionDistance, SymmetricDistance},
+        transformations::polars_test::get_test_data,
+    };
+
+    #[test]
+    fn test_make_count_expr_grouped() -> Fallible<()> {
+        let (lf_domain, lf) = get_test_data()?;
+        let expr_domain = lf_domain.aggregate(["chunk_2_bool"]);
+        let scale: f64 = f64::NAN;
```

Comment:
If it's NAN, does that cause no noise to be applied, so the comparison at the end of this test works?

If that's correct, and it's not a universal convention, it would worry me a little. What if a downstream user has some logic to determine the scale, but there's a soft fail and it ends up NAN, and no noise is applied?

I could be misinterpreting this at a number of points.

Reply:
I'm using the fact that partition lengths are not considered private (information provided by the data curator through the domain margin descriptor) to create a 0-DP mechanism that releases them without noise. The parameter could be anything and no noise will be added.